src/darwin/Framework/CHIP/MTRAsyncWorkQueue.mm

```diff
@@ -29,6 +30,25 @@ typedef NS_ENUM(NSInteger, MTRAsyncWorkItemState) {
     MTRAsyncWorkItemRetryCountBase = MTRAsyncWorkItemRunning, // values >= MTRAsyncWorkItemRunning encode retryCount
 };
 
+// A helper struct that facilitates access to _context while
+//  - only reading the _context weak reference once and retaining a strong
+//    reference for the duration of a particular queue method call
+//  - avoiding calls to `[context description]` under our lock
+struct ContextSnapshot {
```

Comment:
Looking at how this is used, it seems like the pattern is always something like:
````
{
    ContextSnapshot();
    lock();
    // stuff
    unlock();
}
````

That is, the lock is always after the ContextSnapshot constructor and the unlock is always right before the destructor.

Would it make sense to just pass a lock reference to ContextSnapshot and have it do the lock in its ctor and unlock in dtor?

Followup is fine for this change if we do want to do it, just to scope down the changes.

Reply:
I don't really want to turn this into an RAII lock wrapper... If we want to do something like that for our general usage of os_unfair_lock we can do that, but I don't think it should be integrated into this helper.